### PR TITLE
`Communication`: Replace sidebar with new design in messages

### PR DESCRIPTION
**Pull request by egekurt123**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This PR enhances the user experience by updating the sidebar component within the Messages module to a new and improved UI design. By implementing the new UI design, we aim to provide a more intuitive, visually appealing, and user-friendly navigation experience.

Thus, this PR also moves course-wide search filters to its own component.

### Steps for Testing

Prerequisites:
- 1 Instuctor
- 1 Student
- 1 Course with Messages enabled

1. Log in to Artemis with a student account
2. Navigate to Messages
3. Make sure That everything works normally
4. Repeat those steps with an instructor account
5. Also make sure that course-wide search works 

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
| Class/File | Line Coverage 
|------------|--------------:|
| [course-conversations.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6607-38/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/course-conversations.component.ts.html) | 95.54% 
| [accordion-add-options.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6607-38/artifact/shared/Coverage-Report-Client-Tests/app/shared/sidebar/accordion-add-options/accordion-add-options.component.ts.html) | 100% 
| [conversation-options.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6607-38/artifact/shared/Coverage-Report-Client-Tests/app/shared/sidebar/conversation-options/conversation-options.component.ts.html) | 95.12% 
| [sidebar-accordion.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6607-38/artifact/shared/Coverage-Report-Client-Tests/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts.html) | 100% 
| [sidebar-card-small.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6607-38/artifact/shared/Coverage-Report-Client-Tests/app/shared/sidebar/sidebar-card-small/sidebar-card-small.component.ts.html) | 100% 
| [sidebar.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6607-38/artifact/shared/Coverage-Report-Client-Tests/app/shared/sidebar/sidebar.component.ts.html) | 83.33% 
| [course-wide-search.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6607-38/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/course-wide-search/course-wide-search.component.ts.html) | 98.63% 

### Screenshots

Screenshot 2024-06-06 at 21 53 44

Screenshot 2024-06-06 at 21 54 02

### Notes

- Deletion of the legacy Sidebar components will be done in a follow-up PR.
- Integration of course-overview routing will be done in a follow-up PR. Because of the different configuration it needs extensive changes compared to other modules

## Summary by CodeRabbit

- **New Features**
  - Added new form elements for filtering and sorting in the course-wide search, improving user interaction.
  - Introduced video recording capabilities for Playwright tests, enhancing test analysis.

- **Bug Fixes**
  - Updated button and sidebar actions for course messaging to ensure correct functionality.

- **Localization**
  - Added new German localization keys for conversations and messaging features.

- **Tests**
  - Enhanced Playwright test configurations for better performance and reliability.
  - Updated sidebar component tests to include new components and modified navigation paths.

### CODE REVIEW

File: src/main/webapp/app/shared/sidebar/accordion-add-options/accordion-add-options.component.html [hunk 1]
```diff
@@ -0,0 +1,22 @@
+@if (sidebarType === 'conversation' && groupKey !== 'favoriteChannels' && groupKey !== 'hiddenChannels') {
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/accordion-add-options/accordion-add-options.component.html:1]: This Component should be more generic. It is very tailored for the conversations.

File: src/main/webapp/app/shared/sidebar/accordion-add-options/accordion-add-options.component.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        public accountService: AccountService,
+        public conversationService: ConversationService,
+        private courseStorageService: CourseStorageService,
+    ) {}
+
+    ngOnInit() {
+        this.course = this.courseStorageService.getCourse(this.courseId)!;
+    }
+
+    openCreateGroupChatDialog(event: MouseEvent) {
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/accordion-add-options/accordion-add-options.component.ts:48]: All these specific conversation functions should not be inside this generic component

File: src/main/webapp/app/shared/sidebar/conversation-options/conversation-options.component.html [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<div class="option-buttons" (click)="$event.stopPropagation()">
+    <button class="favorite btn btn-sm btn-outline-secondary sidebar-button" (click)="onFavoriteClicked($event)" type="button">
+        <fa-icon [icon]="conversation.isFavorite ? faHeartSolid : faHeartRegular" size="sm" />
+    </button>
+    <div ngbDropdown container="body" class="d-inline-block">
+        <button class="btn btn-sm btn-outline-secondary dropdown-toggle sidebar-button" type="button" ngbDropdownToggle>
+            <fa-icon [icon]="faEllipsisVertical" size="sm" />
+        </button>
+        <div ngbDropdownMenu>
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/conversation-options/conversation-options.component.html:9]: Also this dropdown component should not be tailored to the conversations

File: src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.html [hunk 4]
```diff
@@ -2,14 +2,26 @@
     <div
         id="test-accordion-item-container-{{ i }}"
         class="pointer"
-        [ngClass]="{ 'd-none': !(groupedData[groupKey].entityData | searchFilter: ['title', 'type'] : searchValue)?.length }"
+        [ngClass]="{ 'd-none': sidebarType !== 'conversation' && !(groupedData[groupKey].entityData | searchFilter: ['title', 'type'] : searchValue)?.length }"
     >
         <div id="test-accordion-item-header" class="control-label d-flex align-items-center justify-content-between py-2 bg-module" (click)="toggleGroupCategoryCollapse(groupKey)">
             <div class="my-2">
-                {{ 'artemisApp.courseOverview.sidebar.' + groupKey | artemisTranslate | titlecase }}
+                @if (sidebarType !== 'conversation') {
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.html:9]: Introduce a more generic flat like `showLeadingIcon` you can set its default to `false` and for the conversation data to `true`.

File: src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.html [hunk 5]
```diff
@@ -2,14 +2,26 @@
...
+                    {{ 'artemisApp.courseOverview.sidebar.' + groupKey | artemisTranslate | titlecase }}
+                } @else {
+                    <fa-icon [fixedWidth]="true" [icon]="getIcon(groupKey)" />
+                    {{ 'artemisApp.conversationsLayout.conversationSelectionSideBar.' + groupKey | artemisTranslate }}
+                }
                 ({{ (groupedData[groupKey].entityData | searchFilter: ['title', 'type'] : searchValue)?.length }})
             </div>
-            <div class="pe-3">
+            <div class="icon-container pe-3">
+                <jhi-accordion-add-options
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.html:18]: Same here. We should introduce something like `showMoreOptionsDropdown` and provide the desired data at this point to the new component. This way, we also avoid the add-options-component being specifically designed for the conversations.

File: src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.html [hunk 6]
```diff
@@ -2,14 +2,26 @@
...
-        [ngClass]="{ 'd-none': !(groupedData[groupKey].entityData | searchFilter: ['title', 'type'] : searchValue)?.length }"
+        [ngClass]="{ 'd-none': sidebarType !== 'conversation' && !(groupedData[groupKey].entityData | searchFilter: ['title', 'type'] : searchValue)?.length }"
     >
         <div id="test-accordion-item-header" class="control-label d-flex align-items-center justify-content-between py-2 bg-module" (click)="toggleGroupCategoryCollapse(groupKey)">
             <div class="my-2">
-                {{ 'artemisApp.courseOverview.sidebar.' + groupKey | artemisTranslate | titlecase }}
+                @if (sidebarType !== 'conversation') {
+                    {{ 'artemisApp.courseOverview.sidebar.' + groupKey | artemisTranslate | titlecase }}
+                } @else {
+                    <fa-icon [fixedWidth]="true" [icon]="getIcon(groupKey)" />
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.html:12]: Please avoid funciton calls in the html template

File: src/main/webapp/app/shared/sidebar/sidebar-card-small/sidebar-card-small.component.html [hunk 7]
```diff
@@ -1,9 +1,20 @@
+<!-- Explicitly implemented query parameters for the messages module as a workaround, since routing in the messages module
+ operates differently over the MetisConversations service, it will get adjusted in a followup PR -->
 <div
     id="test-sidebar-card"
     class="pointer rounded-3 col-12 border-start border-5 highlight-card bg-module padding-small"
-    [routerLink]="itemSelected ? '../' + sidebarItem?.id : './' + sidebarItem?.id"
+    [ngClass]="{
+        'border-primary': isConversationUnread
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar-card-small/sidebar-card-small.component.html:7]: Please avoid function calls in the HTML template, this includes getter methods, too.

File: src/main/webapp/app/shared/sidebar/sidebar-card-small/sidebar-card-small.component.html [hunk 8]
```diff
@@ -1,9 +1,20 @@
...
+    [queryParams]="{ conversationId: sidebarItem?.id }"
     (click)="emitStoreLastSelectedItem(sidebarItem.id)"
     [routerLinkActive]="'bg-selected border-selected'"
 >
-    <jhi-sidebar-card-item [sidebarType]="sidebarType" [sidebarItem]="sidebarItem" />
+    <div class="inline-block col-7">
+        <jhi-sidebar-card-item [sidebarType]="sidebarType" [sidebarItem]="sidebarItem" />
+    </div>
+    <div class="inline-block pb-2">
+        <jhi-conversation-options [conversation]="sidebarItem.conversation!" (onUpdateSidebar)="onUpdateSidebar.emit()" />
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar-card-small/sidebar-card-small.component.html:18]: Are you sure that `sidebarItem.conversation` ist definitely defined here?

File: src/main/webapp/app/shared/sidebar/conversation-options/conversation-options.component.html (hunk 3 above)
Comment (egekurt123) [reply src/main/webapp/app/shared/sidebar/conversation-options/conversation-options.component.html:9]: will be solved in a followup PR

**Review by sarpsahinalp — COMMENTED** [review]
Tested locally on ts5, could be unrelated to this PR but I was not able to go to student view or open the messages as an instructor. using the accoung artemis_test_user_16. As a student it works as expected!

**Comment by egekurt123** [comment]
> Tested locally on ts5, could be unrelated to this PR but I was not able to go to student view or open the messages as an instructor using the accoung artemis_test_user_16. As a student it works as expected!

This problem has been occurring since the migration of the test servers. It is because some test users are enrolled as instructors but not as students. You can try testing with test users 1-5 or with your personal account. I have now enrolled test users 15-17 as students on TS5, so testing with these accounts should be possible from now on.

**Review by sarpsahinalp — DISMISSED** [review]
Tested locally works as expected

File: src/main/webapp/app/shared/sidebar/sidebar.component.html [hunk 9]
```diff
@@ -23,6 +23,9 @@
                             [routeParams]="routeParams"
                             [courseId]="courseId"
                             [itemSelected]="itemSelected"
+                            [showAddOptions]="sidebarData.sidebarType === 'conversation'"
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar.component.html:26]: please add an extra key for `showAddOption` default value should be false; and true for conversation
-->      ` [showAddOptions]="sidebarData.showAddOption"`

File: src/main/webapp/app/shared/sidebar/sidebar.component.html [hunk 10]
```diff
@@ -23,6 +23,9 @@
...
+                            [showLeadingIcon]="sidebarData.sidebarType === 'conversation'"
```
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar.component.html:27]: please add an extra key for `showLeadingIcon` default value should be false; and true for conversation
-->  `[showAddOptions]="sidebarData. showLeadingIcon"`

**Comment by rabeatwork** [comment]
When I navigate between the messages and other course pages, this modular window always appears in the messages.
https://github.com/ls1intum/Artemis/assets/75392103/e2c5c565-1c1a-400c-8378-4d650ea9f14a

**Review by Jan-Thurner — APPROVED** [review]
Tested on TS4. Works as expected

**Review by BBesrour — CHANGES_REQUESTED** [review]
Tested on TS2:
- I can't see the `send` button and I don't have the option to scroll down in the right panel. (screenshot 1)
- UI breaks when I go to a channel and send many messages (screenshot 2)

Screenshot 1:
image
Screenshot 2:
image

**Comment by egekurt123** [comment]
@BBesrour Thanks for addressing this issue. Now, in case a user has a small screen, there is another scrollbar that makes it possible to scroll down.

And the Course wide search filters moved to its own component

**Review by BBesrour — APPROVED** [review]
Tested on TS3. Works as expected. Thanks for addressing the issue!

I got a question regarding the search: Should it also search in private messages or only public channels?

**Comment by egekurt123** [comment]
> Tested on TS3. Works as expected. Thanks for addressing the issue!
> 
> I got a question regarding the search: Should it also search in private messages or only public channels?

Currently, it only searches public and private channels, but this PR changes only the UI components of the search.

**Review by az108 — DISMISSED** [review]
Code LGTM and tested on TS3 works as expected and sidebar does not break.

**Review by az108 — DISMISSED** [review]

**Review by Jan-Thurner — DISMISSED** [review]
tested on TS1. Works as expected

**Review by Jan-Thurner — DISMISSED** [review]
reapprove after merge conflict

**Review by az108 — DISMISSED** [review]
reapprove

**Review by BBesrour — APPROVED** [review]
re-approve

**Review by yassinsws — COMMENTED** [review]
Tested on ts4 everything works as expected

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Tested on TS4.

I've noticed a visual bug with longer threads that prevents the whole thread (and/or the text field) from being displayed.

Can you take a look?

TS4 vs. TS1 (I scrolled down as far as I could):

https://github.com/ls1intum/Artemis/assets/38403547/3147611b-70de-4c69-94b5-60b5f0547ede

**Comment by egekurt123** [comment]
> Tested on TS4.
> 
> I've noticed a visual bug with longer threads that prevents the whole thread (and/or the text field) from being displayed.
> 
> Can you take a look?
> 
> TS4 vs. TS1 (I scrolled down as far as I could):
> 
>  2024-06-13.16-15-05.mp4

Yes, there is a minor incompatibility with the scrollbar because it was previously in one div but is now separate (like in every other module that got updated). I have another open PR that updates the UI design of Messages/Threads . Once both PRs are merged, this issue will be resolved. [#8706](https://github.com/ls1intum/Artemis/pull/8706)

**Review by pzdr7 — APPROVED** [review]
Thanks for clarifying - works as expected 👍

**Review by edkaya — APPROVED** [review]
Code

**Review by yassinsws — APPROVED** [review]
reapprove

**Review by asliayk — APPROVED** [review]
tested on ts1, works as expected

**Review by asliayk — DISMISSED** [review]

**Review by BBesrour — DISMISSED** [review]
tested on TS4, lgtm

**Pull request merged**